base_layer/service_framework/src/context/future.rs

```diff
@@ -0,0 +1,136 @@
+/// and which will resolve with the collected `ServiceHandles`.
+pub(crate) fn create_context_notifier_pair(shutdown_signal: ShutdownSignal) -> (Shutdown, ServiceInitializerContext) {
+    let trigger = Shutdown::new();
+    let trigger_signal = trigger.to_signal();
+    (trigger, ServiceInitializerContext::new(shutdown_signal, trigger_signal))
+}
+
+/// Contains context for service initialization.
+///
+/// `ServiceInitializerContext` also implements `Future` and resolves to `ServiceHandles` once
+/// it is signaled to do so.
+#[derive(Clone)]
+pub struct ServiceInitializerContext {
+    inner: FinalServiceHandles,
+    ready_signal: ShutdownSignal,
+}
+
+impl ServiceInitializerContext {
+    /// Create a new ServiceInitializerContext.
+    ///
+    /// ## Parameters
+    /// `shutdown_signal` - signal that is provided to services. If this signal is triggered, services should terminate.
+    /// `ready_signal` - indicates that all services are ready. This should be triggered by the `StackBuilder` once all
+    ///                  initializers have run.
+    pub(crate) fn new(shutdown_signal: ShutdownSignal, ready_signal: ShutdownSignal) -> Self {
+        Self {
+            inner: FinalServiceHandles::new(shutdown_signal),
+            ready_signal,
+        }
+    }
+
+    /// Insert a service handle with the given name
+    pub fn register_handle<H>(&self, handle: H)
+    where H: Any + Send + Sync {
+        self.inner.register(handle);
+    }
+
+    /// Retrieve a handle and downcast it to return type and return a copy, otherwise None is returned
+    pub fn get_handle<H>(&self) -> Option<&H>
+    where H: 'static {
+        self.inner.get_handle()
+    }
+
+    /// Get a handle from the given type (`TypeId`) and downcast it to a type `H`.
+    /// If the item does not exist or the downcast fails, a panic occurs
+    pub fn expect_handle<H>(&self) -> H
+    where H: Clone + 'static {
+        self.inner.expect_handle()
+    }
+
+    /// Call the given function with the final handles once this future is ready (`notify_ready` is called).
+    pub fn lazy_service<F, S>(&self, service_fn: F) -> LazyService<F, Self, S>
+    where F: FnOnce(FinalServiceHandles) -> S {
+        LazyService::new(self.clone(), service_fn)
+    }
+
+    /// Spawn a task once handles are ready. The resolved handles are passed into this closure.
+    pub fn spawn_when_ready<F, Fut>(self, f: F) -> task::JoinHandle<Fut::Output>
+    where
+        F: FnOnce(FinalServiceHandles) -> Fut + Send + 'static,
+        Fut: Future<Output = ()> + Send + 'static,
+    {
+        task::spawn(self.wait_ready().then(f))
+    }
+
+    pub async fn wait_ready(self) -> FinalServiceHandles {
+        let _ = self.ready_signal.await;
+        self.inner
+    }
+
+    /// Spawns the task immediately without waiting for all initializers to complete
+    pub fn spawn_immediately<T>(&self, task: T) -> task::JoinHandle<T::Output>
```

Comment:
Why would a caller use this instead of just calling `task::spawn()` directly? The method seems unnecessary and could lead to dead code

Reply:
The reasons could be:
1. You don't need to have tokio as a dependency or import it (meh - it's fine I guess)
1. If we want to change the way services spawn, the change happens here (probably not going to happen)
1. We use `spawn_when_ready` to spawn tasks in the context, so this is a counterpart which doesn't wait until ready (kind of pedantic)

I think what I'll do is remove this and change the caller to just spawn when ready (even though it doesn't depend on any handles) - there is practically no difference.